### PR TITLE
Self relation Module associations (task #10059)

### PR DESCRIPTION
**Pull request by eliamariutti**
Add support to M:M associations with the same table (self-referencing).

### CODE REVIEW

File: src/Model/AssociationsAwareTrait.php [hunk 1]
```diff
@@ -121,10 +121,35 @@ private function setByTypeRelation(string $module, array $fields) : void
         }
 
         foreach ($fields as $field) {
-            $this->setByTypeRelationField($module, $field, $moduleField);
+            $this->setByTypeRelationField($module, $field, $moduleField, $this->isSelfRelated($fields));
```
Comment (georgeconstantinou) [inline src/Model/AssociationsAwareTrait.php]: @eliamariutti you can call `$this->isSelfRelated($fields)` before the _foreach_ loop, no need to be called for each field.

File: src/Model/AssociationsAwareTrait.php [hunk 2]
```diff
@@ -120,36 +120,56 @@ private function setByTypeRelation(string $module, array $fields) : void
...
             $duplicate = $first;
         }
 
-        return false;
+        if (count($selfrelated) === 2) {
+            return $selfrelated;
+        }
+
+        if (count($selfrelated) > 2) {
+            throw new Exception('Related table with more than two fields');
```
Comment (georgeconstantinou) [inline src/Model/AssociationsAwareTrait.php]: @eliamariutti can we change the message here to: _Many-to-many self-association with more than two fields is not supported._

Also is better to throw a `\RuntimeException` instead.

File: src/Model/AssociationsAwareTrait.php [hunk 3]
```diff
@@ -120,11 +120,58 @@ private function setByTypeRelation(string $module, array $fields) : void
...
+    /**
+     * Check if the relation module is "self-related" : is true when
+     * two related fields point to the some module. In this case will
+     * be return an array with two CsvField object.
+     *
+     * @param  mixed[] $fields Module fields
+     * @return CsvMigrations\FieldHandlers\CsvField[]|null
+     * @throws Exception in case there are more than two fields pointing the some module.
+     */
+    private function isSelfRelated(array $fields)
```
Comment (softius) [inline src/Model/AssociationsAwareTrait.php]: @eliamariutti @georgeconstantinou functions with `is` or `has` prefix is better to return boolean values. 

PS. dont forget the tests

Comment (georgeconstantinou) [reply src/Model/AssociationsAwareTrait.php]: @softius true, @eliamariutti I think we need to change the method name to `getSelfRelated`.

File: src/Model/AssociationsAwareTrait.php [hunk 4]
```diff
@@ -120,11 +121,57 @@ private function setByTypeRelation(string $module, array $fields) : void
...
+        }
+
+        if (count($selfrelated) === 2) {
+            return $selfrelated;
+        }
+
+        if (count($selfrelated) > 2) {
+            throw new RuntimeException('Many-to-many self-association with more than two fields is not supported.');
+        }
+    }
```
Comment (georgeconstantinou) [inline src/Model/AssociationsAwareTrait.php]: @eliamariutti I think here you can return an empty array, for cases where the `selftrelated` count is less than 2. This way you can remove the nullable from the return type-hint and the void from the return doc-block.

**Pull request merged**